src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs

```diff
@@ -451,6 +451,12 @@ private BoundForEachStatement BindForEachPartsWorker(BindingDiagnosticBag diagno
             CompoundUseSiteInfo<AssemblySymbol> useSiteInfo = GetNewCompoundUseSiteInfo(diagnostics);
             Conversion elementConversionClassification = this.Conversions.ClassifyConversionFromType(inferredType.Type, iterationVariableType.Type, isChecked: CheckOverflowAtRuntime, ref useSiteInfo, forCast: true);
 
+            if (elementConversionClassification.Kind != ConversionKind.Identity && (IterationVariable.RefKind == RefKind.Ref || IterationVariable.RefKind == RefKind.RefReadOnly) && builder.CurrentPropertyGetter.RefKind == RefKind.Ref)
```

Comment:
I think all we will need is

```suggestion
            if (elementConversionClassification.Kind != ConversionKind.Identity && IterationVariable.RefKind is RefKind.Ref or RefKind.RefReadOnly)
```

Reply:
Verified through added test